### Import Libary

In [339]:
import pandas as pd
import os

### settings

In [60]:
# allows more rows columns to be displayed
pd.set_option('display.max_rows', 200)
pd.set_option('display.max_columns', 200)

### Read in files

In [5]:
# list out all the csv files
path = './DataAnalystAssignment/'
csv_files = [f for f in os.listdir(path) if 'csv' in f]
csv_files

['circuits.csv',
 'constructors.csv',
 'constructor_standings.csv',
 'drivers.csv',
 'driverStandings.csv',
 'driver_standings.csv',
 'lapTimes.csv',
 'pitStops.csv',
 'races.csv',
 'results.csv',
 'status.csv']

Looks like there are 2 'driver standing' files, this is different to what was specified

In [6]:
circuits = pd.read_csv(path + 'circuits.csv', encoding = "ISO-8859-1")
constructor_standings = pd.read_csv(path + 'constructor_standings' + '.csv', encoding = "ISO-8859-1")
constructors = pd.read_csv(path + 'constructors' + '.csv', encoding = "ISO-8859-1")
driverStandings = pd.read_csv(path + 'driverStandings' + '.csv', encoding = "ISO-8859-1")
driver_standings = pd.read_csv(path + 'driver_standings' + '.csv', encoding = "ISO-8859-1")
drivers = pd.read_csv(path + 'drivers' + '.csv', encoding = "ISO-8859-1")
lapTimes = pd.read_csv(path + 'lapTimes' + '.csv', encoding = "ISO-8859-1")
pitStops = pd.read_csv(path + 'pitStops' + '.csv', encoding = "ISO-8859-1")
races = pd.read_csv(path + 'races' + '.csv', encoding = "ISO-8859-1")
results = pd.read_csv(path + 'results' + '.csv', encoding = "ISO-8859-1")
status = pd.read_csv(path + 'status' + '.csv', encoding = "ISO-8859-1")

### Data Exploration and Cleaning

#### the 2 driver standing csv files

In [25]:
# check size 
driver_standings.shape, driverStandings.shape

((32924, 7), (31726, 7))

In [29]:
driverStandings.describe()

driverStandingsId        raceId      driverId        points  \
count       31726.000000  31726.000000  31726.000000  31726.000000   
mean        39545.294806    537.923438    280.237944     10.623804   
std         21234.446977    261.533662    241.105400     28.818181   
min             1.000000      1.000000      1.000000      0.000000   
25%         18503.250000    329.000000     84.000000      0.000000   
50%         47758.500000    566.000000    205.000000      0.000000   
75%         56741.750000    769.000000    437.000000      8.000000   
max         68608.000000    988.000000    843.000000    397.000000   

           position          wins  
count  31726.000000  31726.000000  
mean      20.571361      0.252002  
std       16.728735      0.929208  
min        1.000000      0.000000  
25%        9.000000      0.000000  
50%       17.000000      0.000000  
75%       27.000000      0.000000  
max      108.000000     13.000000

In [30]:
driver_standings.describe()

driverStandingsId        raceId      driverId        points  \
count       32924.000000  32924.000000  32924.000000  32924.000000   
mean        40632.826418    555.404933    293.636830     12.135245   
std         21582.983395    272.056972    254.712819     32.544135   
min             1.000000      1.000000      1.000000      0.000000   
25%         18943.750000    337.000000     84.000000      0.000000   
50%         48668.500000    581.000000    209.000000      1.000000   
75%         57640.250000    783.000000    477.000000      9.000000   
max         70209.000000   1047.000000    851.000000    413.000000   

           position          wins  
count  32924.000000  32924.000000  
mean      20.210819      0.261511  
std       16.563971      0.961874  
min        1.000000      0.000000  
25%        8.000000      0.000000  
50%       16.000000      0.000000  
75%       26.000000      0.000000  
max      108.000000     13.000000

In [9]:
# column names
driverStandings.columns, driver_standings.columns

(Index(['driverStandingsId', 'raceId', 'driverId', 'points', 'position',
        'positionText', 'wins'],
       dtype='object'),
 Index(['driverStandingsId', 'raceId', 'driverId', 'points', 'position',
        'positionText', 'wins'],
       dtype='object'))

- looks like the same table, but driverStandings missing about 1200 rows compared to driverStandings
- let's check if there are any data differences

In [24]:
# by applying outer join, check which table has rows with values that the other table does not have
diff_driverStanding = driver_standings.merge(driverStandings,indicator = True, how='outer').loc[lambda x : x['_merge']!='both']
diff_driverStanding['_merge'].unique()

[left_only]
Categories (1, object): [left_only]

Result shows 'left _only', hence the right df (driverStandings) does not have any data that does not already exist in the left df (driver_standings). <br>
Hence:

**Assumption**: use the sheet with more data: driver_standings

In [26]:
driverStandings.describe()

driverStandingsId        raceId      driverId        points  \
count       31726.000000  31726.000000  31726.000000  31726.000000   
mean        39545.294806    537.923438    280.237944     10.623804   
std         21234.446977    261.533662    241.105400     28.818181   
min             1.000000      1.000000      1.000000      0.000000   
25%         18503.250000    329.000000     84.000000      0.000000   
50%         47758.500000    566.000000    205.000000      0.000000   
75%         56741.750000    769.000000    437.000000      8.000000   
max         68608.000000    988.000000    843.000000    397.000000   

           position          wins  
count  31726.000000  31726.000000  
mean      20.571361      0.252002  
std       16.728735      0.929208  
min        1.000000      0.000000  
25%        9.000000      0.000000  
50%       17.000000      0.000000  
75%       27.000000      0.000000  
max      108.000000     13.000000

In [146]:
# check for nan
driverStandings.isna().sum()

driverStandingsId    0
raceId               0
driverId             0
points               0
position             0
positionText         0
wins                 0
dtype: int64

#### results.csv

In [344]:
results.head(10)

resultId  raceId  driverId  constructorId  number  grid  position  \
0         1      18         1              1    22.0     1       1.0   
1         2      18         2              2     3.0     5       2.0   
2         3      18         3              3     7.0     7       3.0   
3         4      18         4              4     5.0    11       4.0   
4         5      18         5              1    23.0     3       5.0   
5         6      18         6              3     8.0    13       6.0   
6         7      18         7              5    14.0    17       7.0   
7         8      18         8              6     1.0    15       8.0   
8         9      18         9              2     4.0     2       NaN   
9        10      18        10              7    12.0    18       NaN   

  positionText  positionOrder  points  laps     time  milliseconds  \
0            1              1    10.0    58  34:50.6     5690616.0   
1            2              2     8.0    58    5.478     5696094.0   
2            3              3     6.0    58    8.163     5698779.0   
3            4              4     5.0    58   17.181     5707797.0   
4            5              5     4.0    58   18.014     5708630.0   
5            6              6     3.0    57      NaN           NaN   
6            7              7     2.0    55      NaN           NaN   
7            8              8     1.0    53      NaN           NaN   
8            R              9     0.0    47      NaN           NaN   
9            R             10     0.0    43      NaN           NaN   

   fastestLap  rank fastestLapTime fastestLapSpeed  statusId  
0        39.0   2.0        01:27.5           218.3         1  
1        41.0   3.0        01:27.7         217.586         1  
2        41.0   5.0        01:28.1         216.719         1  
3        58.0   7.0        01:28.6         215.464         1  
4        43.0   1.0        01:27.4         218.385         1  
5        50.0  14.0        01:29.6         212.974        11  
6        22.0  12.0        01:29.5         213.224         5  
7        20.0   4.0        01:27.9          217.18         5  
8        15.0   9.0        01:28.8           215.1         4  
9        23.0  13.0        01:29.6         213.166         3

In [31]:
results.describe()

resultId        raceId      driverId  constructorId        number  \
count  23777.000000  23777.000000  23777.000000   23777.000000  23771.000000   
mean   11889.481053    487.203937    226.515961      46.281785     16.965462   
std     6864.691322    269.904857    231.386102      56.174091     13.644798   
min        1.000000      1.000000      1.000000       1.000000      0.000000   
25%     5945.000000    273.000000     55.000000       6.000000      7.000000   
50%    11889.000000    478.000000    154.000000      25.000000     15.000000   
75%    17833.000000    718.000000    314.000000      57.000000     23.000000   
max    23781.000000    988.000000    843.000000     210.000000    208.000000   

               grid      position  positionOrder        points          laps  \
count  23777.000000  13227.000000   23777.000000  23777.000000  23777.000000   
mean      11.270303      7.782264      13.081591      1.601403     45.270598   
std        7.346436      4.745105       7.824711      3.665154     30.525404   
min        0.000000      1.000000       1.000000      0.000000      0.000000   
25%        5.000000      4.000000       7.000000      0.000000     20.000000   
50%       11.000000      7.000000      13.000000      0.000000     52.000000   
75%       17.000000     11.000000      19.000000      1.000000     66.000000   
max       34.000000     33.000000      39.000000     50.000000    200.000000   

       milliseconds   fastestLap         rank      statusId  
count  6.003000e+03  5383.000000  5531.000000  23777.000000  
mean   6.303313e+06    41.061676    10.598807     18.242293  
std    1.721748e+06    17.156435     6.272457     26.380824  
min    1.474899e+06     2.000000     0.000000      1.000000  
25%    5.442948e+06    29.000000     5.000000      1.000000  
50%    5.859428e+06    44.000000    11.000000     11.000000  
75%    6.495440e+06    53.000000    16.000000     16.000000  
max    1.509054e+07    78.000000    24.000000    136.000000

In [340]:
# the max value for points is 50, but according to the rules 1st place gets 25 points, look further
results[results['points'] == 50]

resultId  raceId  driverId  constructorId  number  grid  position  \
22514     22518     918         1            131    44.0     2       1.0   

      positionText  positionOrder  points  laps     time  milliseconds  \
22514            1              1    50.0    55  39:02.6     5942619.0   

       fastestLap  rank fastestLapTime fastestLapSpeed  statusId  
22514        49.0   4.0        01:45.6         189.342         1

In [341]:
# check this particular raceId
races[races['raceId'] == 918]

raceId  year  round  circuitId                  name        date  \
915     918  2014     19         24  Abu Dhabi Grand Prix  2014-11-23   

         time                                                url  
915  13:00:00  http://en.wikipedia.org/wiki/2014_Abu_Dhabi_Gr...

According to google, this was the only time up to present where double points was given.

In [34]:
# check if any nan exist
results.isnull().sum()

resultId               0
raceId                 0
driverId               0
constructorId          0
number                 6
grid                   0
position           10550
positionText           0
positionOrder          0
points                 0
laps                   0
time               17773
milliseconds       17774
fastestLap         18394
rank               18246
fastestLapTime     18394
fastestLapSpeed    18394
statusId               0
dtype: int64

lets look at those columns with nan

In [39]:
# position has nan, but both positionText and positionOrder doesn't
results[['position', 'positionText', 'positionOrder']][results['position'].isna()]

position positionText  positionOrder
8           NaN            R              9
9           NaN            R             10
10          NaN            R             11
11          NaN            R             12
12          NaN            R             13
...         ...          ...            ...
23754       NaN            R             18
23755       NaN            R             19
23756       NaN            R             20
23775       NaN            R             19
23776       NaN            R             20

[10550 rows x 3 columns]

In [79]:
# make a copy of the original, and check if all the position columns always match
results_analysis = results.copy(deep=True)
results_analysis['match'] = results_analysis.apply(lambda x : 'match' if x['position'] == float(x['positionOrder']) else "no match", axis=1)

In [80]:
results_analysis['position'][results_analysis['match'] == 'no match'].unique()

array([nan,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10., 11., 12., 13., 14.,
       15.])

In [84]:
# check non nan and not matching rows 
results_analysis[(results_analysis['match'] == 'no match') & (results_analysis['position'].notnull())]

resultId  raceId  driverId  constructorId  number  grid  position  \
10656     10657     452       182              1     8.0     9       3.0   
10657     10658     452       172              3     2.0    18       4.0   
10658     10659     452       175              6    27.0     3       5.0   
10659     10660     452       176             21    29.0    20       6.0   
10660     10661     452       117              4    15.0     2       7.0   
10661     10662     452       118             53    35.0     5       8.0   
10662     10663     452       192             21    30.0    23       9.0   
10663     10664     452       163              6    28.0     6      10.0   
10664     10665     452       193             25     4.0    21      11.0   
10665     10666     452        95             32    12.0    22      12.0   
10666     10667     452       188             55    34.0    19      13.0   
10667     10668     452       194             50    17.0    26      14.0   
10668     10669     452       181             54     9.0    25      15.0   

      positionText  positionOrder  points  laps       time  milliseconds  \
10656            3              2     4.0    63     51.883     6559614.0   
10657            4              3     3.0    63  +1:13.951     6581682.0   
10658            5              4     2.0    63  +1:18.117     6585848.0   
10659            6              5     1.0    63  +1:18.207     6585938.0   
10660            7              6     0.0    62        NaN           NaN   
10661            8              7     0.0    62        NaN           NaN   
10662            9              8     0.0    62        NaN           NaN   
10663           10              9     0.0    62        NaN           NaN   
10664           11             10     0.0    62        NaN           NaN   
10665           12             11     0.0    61        NaN           NaN   
10666           13             12     0.0    60        NaN           NaN   
10667           14             13     0.0    59        NaN           NaN   
10668           15             14     0.0    59        NaN           NaN   

       fastestLap  rank fastestLapTime fastestLapSpeed  statusId     match  
10656         NaN   NaN            NaN             NaN         1  no match  
10657         NaN   NaN            NaN             NaN         1  no match  
10658         NaN   NaN            NaN             NaN         1  no match  
10659         NaN   NaN            NaN             NaN         1  no match  
10660         NaN   NaN            NaN             NaN        11  no match  
10661         NaN   NaN            NaN             NaN        11  no match  
10662         NaN   NaN            NaN             NaN        11  no match  
10663         NaN   NaN            NaN             NaN        11  no match  
10664         NaN   NaN            NaN             NaN        11  no match  
10665         NaN   NaN            NaN             NaN        12  no match  
10666         NaN   NaN            NaN             NaN        13  no match  
10667         NaN   NaN            NaN             NaN        14  no match  
10668         NaN   NaN            NaN             NaN        14  no match

seems like all from the same raceId: 452. Not sure why these rows are not matching, every other row has been matching except for when nan exists. <br>
**Assumption**: 
Given both position and positionText have the same value, assume the error is with positionOrder. 

In [387]:
# fix positionOrder data
resultIds= results_analysis['resultId'][(results_analysis['match'] == 'no match') & (
    results_analysis['position'].notnull())].tolist()

In [409]:
# update positionOrder according to position
for ids in resultIds:
    results_analysis.loc[results_analysis['resultId'] == ids, 'positionOrder'] = int(
        results_analysis[results_analysis['resultId'] == ids].iloc[0]['position'])

In [411]:
results_analysis[(results_analysis['match'] == 'no match') & (results_analysis['position'].notnull())]

resultId  raceId  driverId  constructorId  number  grid  position  \
10656     10657     452       182              1     8.0     9       3.0   
10657     10658     452       172              3     2.0    18       4.0   
10658     10659     452       175              6    27.0     3       5.0   
10659     10660     452       176             21    29.0    20       6.0   
10660     10661     452       117              4    15.0     2       7.0   
10661     10662     452       118             53    35.0     5       8.0   
10662     10663     452       192             21    30.0    23       9.0   
10663     10664     452       163              6    28.0     6      10.0   
10664     10665     452       193             25     4.0    21      11.0   
10665     10666     452        95             32    12.0    22      12.0   
10666     10667     452       188             55    34.0    19      13.0   
10667     10668     452       194             50    17.0    26      14.0   
10668     10669     452       181             54     9.0    25      15.0   

      positionText  positionOrder  points  laps       time  milliseconds  \
10656            3              3     4.0    63     51.883     6559614.0   
10657            4              4     3.0    63  +1:13.951     6581682.0   
10658            5              5     2.0    63  +1:18.117     6585848.0   
10659            6              6     1.0    63  +1:18.207     6585938.0   
10660            7              7     0.0    62        NaN           NaN   
10661            8              8     0.0    62        NaN           NaN   
10662            9              9     0.0    62        NaN           NaN   
10663           10             10     0.0    62        NaN           NaN   
10664           11             11     0.0    62        NaN           NaN   
10665           12             12     0.0    61        NaN           NaN   
10666           13             13     0.0    60        NaN           NaN   
10667           14             14     0.0    59        NaN           NaN   
10668           15             15     0.0    59        NaN           NaN   

       fastestLap  rank fastestLapTime fastestLapSpeed  statusId     match  
10656         NaN   NaN            NaN             NaN         1  no match  
10657         NaN   NaN            NaN             NaN         1  no match  
10658         NaN   NaN            NaN             NaN         1  no match  
10659         NaN   NaN            NaN             NaN         1  no match  
10660         NaN   NaN            NaN             NaN        11  no match  
10661         NaN   NaN            NaN             NaN        11  no match  
10662         NaN   NaN            NaN             NaN        11  no match  
10663         NaN   NaN            NaN             NaN        11  no match  
10664         NaN   NaN            NaN             NaN        11  no match  
10665         NaN   NaN            NaN             NaN        12  no match  
10666         NaN   NaN            NaN             NaN        13  no match  
10667         NaN   NaN            NaN             NaN        14  no match  
10668         NaN   NaN            NaN             NaN        14  no match

check the other nan value in: fastestLap, rank, fastestLapTime and fastestLapSpeed, these could affect task 4 (fastest race tracks based on lap times)

In [101]:
results['positionText'][results['fastestLap'].isna()].unique()

array(['R', 'W', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11',
       '12', '13', '14', '15', '16', '17', 'D', 'F', '18', 'E', 'N', '19',
       '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30',
       '31', '32', '33'], dtype=object)

In [431]:
# check if there are missing times for position at 1
results[['raceId', 'driverId','positionText']][(results['fastestLap'].isna()) & (results['positionText'] == '1')]

raceId  driverId positionText
1874      108        14            1
1894      109         8            1
1914      110        21            1
1934      111        30            1
1954      112        30            1
...       ...       ...          ...
20185     828       579            1
20266     784       579            1
20288     780       475            1
20297     728       373            1
20822     843         1            1

[717 rows x 3 columns]

In [432]:
# check a random raceId
results[['raceId', 'driverId', 'position', 'positionText', 'fastestLapTime']][(results['raceId']==108)]

raceId  driverId  position positionText fastestLapTime
1874     108        14       1.0            1            NaN
1875     108        31       2.0            2            NaN
1876     108         8       3.0            3            NaN
1877     108        30       4.0            4            NaN
1878     108        15       5.0            5            NaN
1879     108        49       6.0            6            NaN
1880     108         4       7.0            7            NaN
1881     108        23       8.0            8            NaN
1882     108        35       9.0            9            NaN
1883     108        18      10.0           10            NaN
1884     108        50      11.0           11            NaN
1885     108        21       NaN            R            NaN
1886     108        42       NaN            R            NaN
1887     108        44       NaN            R            NaN
1888     108         2       NaN            R            NaN
1889     108        51       NaN            R            NaN
1890     108        17       NaN            R            NaN
1891     108        43       NaN            R            NaN
1892     108        52       NaN            R            NaN
1893     108        22       NaN            R            NaN

In [164]:
# for the same raceId, check on races for the driverId at position 1
lapTimes[(lapTimes['raceId'] == 108) & (lapTimes['driverId'] == 14)].sort_values('time')

raceId  driverId  lap  position      time  milliseconds
230300     108        14   28         3  1:28.272         88272
230301     108        14   29         2  1:28.440         88440
230298     108        14   26         3  1:28.499         88499
230303     108        14   31         2  1:28.582         88582
230297     108        14   25         3  1:28.591         88591
230296     108        14   24         3  1:28.604         88604
230299     108        14   27         3  1:28.622         88622
230302     108        14   30         2  1:28.650         88650
230325     108        14   53         1  1:29.006         89006
230295     108        14   23         3  1:29.009         89009
230326     108        14   54         1  1:29.063         89063
230294     108        14   22         3  1:29.102         89102
230322     108        14   50         1  1:29.124         89124
230321     108        14   49         1  1:29.224         89224
230324     108        14   52         1  1:29.246         89246
230323     108        14   51         1  1:29.298         89298
230318     108        14   46         2  1:29.344         89344
230315     108        14   43         3  1:29.361         89361
230319     108        14   47         2  1:29.361         89361
230317     108        14   45         3  1:29.365         89365
230316     108        14   44         3  1:29.371         89371
230287     108        14   15         7  1:29.419         89419
230327     108        14   55         1  1:29.426         89426
230314     108        14   42         3  1:29.487         89487
230311     108        14   39         3  1:29.580         89580
230308     108        14   36         4  1:29.654         89654
230320     108        14   48         1  1:29.761         89761
230306     108        14   34         5  1:29.852         89852
230307     108        14   35         5  1:29.853         89853
230288     108        14   16         6  1:29.922         89922
230328     108        14   56         1  1:29.923         89923
230312     108        14   40         3  1:29.970         89970
230309     108        14   37         4  1:30.134         90134
230313     108        14   41         3  1:30.177         90177
230329     108        14   57         1  1:30.289         90289
230285     108        14   13         7  1:30.469         90469
230286     108        14   14         7  1:30.483         90483
230293     108        14   21         3  1:30.836         90836
230289     108        14   17         3  1:31.007         91007
230310     108        14   38         4  1:31.087         91087
230330     108        14   58         1  1:31.285         91285
230278     108        14    6        15  1:32.569         92569
230284     108        14   12         7  1:32.827         92827
230277     108        14    5        17  1:33.070         93070
230279     108        14    7        12  1:33.637         93637
230305     108        14   33         5  1:34.559         94559
230276     108        14    4        18  1:36.308         96308
230280     108        14    8         8  1:37.437         97437
230275     108        14    3        20  1:42.318        102318
230304     108        14   32         5  1:55.302        115302
230273     108        14    1        10  1:56.961        116961
230274     108        14    2        20  2:06.390        126390
230281     108        14    9         8  2:11.164        131164
230292     108        14   20         3  2:25.995        145995
230283     108        14   11         7  2:27.526        147526
230290     108        14   18         3  2:31.400        151400
230291     108        14   19         3  2:35.654        155654
230282     108        14   10         7  2:38.259        158259

In [168]:
lapTimes.groupby(['raceId', 'driverId'])['time'].agg('min').reset_index()[ \
    lapTimes.groupby(['raceId', 'driverId'])['time'].agg('min').reset_index()['raceId']==108]

raceId  driverId      time
2118     108         2  1:33.519
2119     108         4  1:28.170
2120     108         8  1:27.724
2121     108        14  1:28.272
2122     108        15  1:28.638
2123     108        17  1:29.697
2124     108        18  1:28.600
2125     108        21  1:29.274
2126     108        22  1:37.086
2127     108        23  1:28.617
2128     108        30  1:27.759
2129     108        31  1:27.942
2130     108        35  1:28.770
2131     108        42  1:29.217
2132     108        43  1:33.753
2133     108        44  1:29.694
2134     108        49  1:29.096
2135     108        50  1:31.785
2136     108        51  1:33.139
2137     108        52  1:36.644

looks like the missing times can be filled in from lapTimes.csv

In [423]:
# first find the fasterest time and lap for each raceId and driverId from lapTimes
lalapTimes_agg = lapTimes.copy(deep=True)
lalapTimes_agg = lalapTimes_agg.groupby(['raceId', 'driverId'])['time'].agg('min').reset_index()
lalapTimes_agg = lalapTimes_agg.rename(columns={'time':'fastestLapTime'})

In [425]:
# combine_first() will override nan values at the same position if the other df has a non nan value at the same position
results_filllaptime = lalapTimes_agg.combine_first(results_analysis)
# drop 'match' column generated from previously
del results_filllaptime['match']

In [427]:
# some sanity check
print(results_filllaptime['fastestLapTime'][(results_filllaptime['raceId']==108) & (results_filllaptime['driverId']==2)].iloc[0] == \
lalapTimes_agg['fastestLapTime'][(lalapTimes_agg['raceId']==108) & (results_filllaptime['driverId']==2)].iloc[0])

print(results_filllaptime['fastestLapTime'][(results_filllaptime['raceId']==1) & (results_filllaptime['driverId']==2)].iloc[0] == \
lalapTimes_agg['fastestLapTime'][(lalapTimes_agg['raceId']==1) & (results_filllaptime['driverId']==2)].iloc[0])

True
True


In [428]:
# can't do much about the rest
results_filllaptime.isna().sum()

constructorId          0
driverId               0
fastestLap         18394
fastestLapSpeed    18394
fastestLapTime     12338
grid                   0
laps                   0
milliseconds       17774
number                 6
points                 0
position           10550
positionOrder          0
positionText           0
raceId                 0
rank               18246
resultId               0
statusId               0
time               17773
dtype: int64

In [429]:
columns

['resultId',
 'raceId',
 'driverId',
 'constructorId',
 'number',
 'grid',
 'position',
 'positionText',
 'positionOrder',
 'points',
 'laps',
 'time',
 'milliseconds',
 'fastestLap',
 'rank',
 'fastestLapTime',
 'fastestLapSpeed',
 'statusId']

In [430]:
# let's keep the column in the same order as before
columns = results.columns.tolist()
results_filllaptime = results_filllaptime.reindex(columns=columns)
# output to file
results_filllaptime.to_csv('./WrangledDatasets/results_wrangled.csv', index=False)

#### races.csv

In [295]:
# look at a general statistic descrbe, see if anything that would not make sense, for example having min year earlier than 1950
races.describe()

raceId         year       round   circuitId
count   997.000000   997.000000  997.000000  997.000000
mean    500.034102  1988.990973    8.233701   21.762287
std     289.590656    18.906633    4.884669   16.437073
min       1.000000  1950.000000    1.000000    1.000000
25%     250.000000  1974.000000    4.000000    9.000000
50%     499.000000  1990.000000    8.000000   18.000000
75%     748.000000  2005.000000   12.000000   30.000000
max    1009.000000  2018.000000   21.000000   73.000000

In [119]:
# check if any nan exists
races.isna().sum()

raceId         0
year           0
round          0
circuitId      0
name           0
date           0
time         731
url            0
dtype: int64

In [287]:
# check the nan values
races[races['time'].isna()].head(20)

raceId  year  round  circuitId                      name        date  \
89       90  2004      1          1     Australian Grand Prix  2004-03-07   
90       91  2004      2          2      Malaysian Grand Prix  2004-03-21   
91       92  2004      3          3        Bahrain Grand Prix  2004-04-04   
92       93  2004      4         21     San Marino Grand Prix  2004-04-25   
93       94  2004      5          4        Spanish Grand Prix  2004-05-09   
94       95  2004      6          6         Monaco Grand Prix  2004-05-23   
95       96  2004      7         20       European Grand Prix  2004-05-30   
96       97  2004      8          7       Canadian Grand Prix  2004-06-13   
97       98  2004      9         19  United States Grand Prix  2004-06-20   
98       99  2004     10          8         French Grand Prix  2004-07-04   
99      100  2004     11          9        British Grand Prix  2004-07-11   
100     101  2004     12         10         German Grand Prix  2004-07-25   
101     102  2004     13         11      Hungarian Grand Prix  2004-08-15   
102     103  2004     14         13        Belgian Grand Prix  2004-08-29   
103     104  2004     15         14        Italian Grand Prix  2004-09-12   
104     105  2004     16         17        Chinese Grand Prix  2004-09-26   
105     106  2004     17         22       Japanese Grand Prix  2004-10-10   
106     107  2004     18         18      Brazilian Grand Prix  2004-10-24   
107     108  2003      1          1     Australian Grand Prix  2003-03-09   
108     109  2003      2          2      Malaysian Grand Prix  2003-03-23   

    time                                                url  
89   NaN  http://en.wikipedia.org/wiki/2004_Australian_G...  
90   NaN  http://en.wikipedia.org/wiki/2004_Malaysian_Gr...  
91   NaN  http://en.wikipedia.org/wiki/2004_Bahrain_Gran...  
92   NaN  http://en.wikipedia.org/wiki/2004_San_Marino_G...  
93   NaN  http://en.wikipedia.org/wiki/2004_Spanish_Gran...  
94   NaN  http://en.wikipedia.org/wiki/2004_Monaco_Grand...  
95   NaN  http://en.wikipedia.org/wiki/2004_European_Gra...  
96   NaN  http://en.wikipedia.org/wiki/2004_Canadian_Gra...  
97   NaN  http://en.wikipedia.org/wiki/2004_United_State...  
98   NaN  http://en.wikipedia.org/wiki/2004_French_Grand...  
99   NaN  http://en.wikipedia.org/wiki/2004_British_Gran...  
100  NaN  http://en.wikipedia.org/wiki/2004_German_Grand...  
101  NaN  http://en.wikipedia.org/wiki/2004_Hungarian_Gr...  
102  NaN  http://en.wikipedia.org/wiki/2004_Belgian_Gran...  
103  NaN  http://en.wikipedia.org/wiki/2004_Italian_Gran...  
104  NaN  http://en.wikipedia.org/wiki/2004_Chinese_Gran...  
105  NaN  http://en.wikipedia.org/wiki/2004_Japanese_Gra...  
106  NaN  http://en.wikipedia.org/wiki/2004_Brazilian_Gr...  
107  NaN  http://en.wikipedia.org/wiki/2003_Australian_G...  
108  NaN  http://en.wikipedia.org/wiki/2003_Malaysian_Gr...

In [294]:
races['time'].unique()

array(['06:00:00', '09:00:00', '07:00:00', '12:00:00', '05:00:00',
       '16:00:00', '11:00:00', '04:30:00', '11:30:00', '17:00:00',
       '03:00:00', '14:30:00', '15:00:00', '14:00:00', '13:00:00', nan,
       '08:00:00', '09:30:00', '18:00:00', '19:00:00', '20:00:00'],
      dtype=object)

races.csv looks fine, the 731 nan rows are just the time the races were held, this should not have affect any analysis as per the tasks. Can be ignored at this stage.

#### drivers.csv

In [120]:
drivers.isna().sum()

driverId         0
driverRef        0
number         804
code           757
forename         0
surname          0
dob              1
nationality      0
url              1
dtype: int64

In [224]:
drivers.describe()

driverId     number
count  842.000000  38.000000
mean   421.540380  30.500000
std    243.275985  27.682319
min      1.000000   2.000000
25%    211.250000  10.250000
50%    421.500000  21.500000
75%    631.750000  38.250000
max    843.000000  99.000000

drivers.csv looks fine, the 2 nan rows should not affect any analysis as per the tasks. Can be ignored at this stage.

#### constructor_standings

In [121]:
# check if any nan
constructor_standings.isna().sum()

constructorStandingsId    0
raceId                    0
constructorId             0
points                    0
position                  0
positionText              0
wins                      0
dtype: int64

In [227]:
# raceId max value is higher than raceId max value from race.csv ()
constructor_standings.describe()

constructorStandingsId        raceId  constructorId        points  \
count            12486.000000  12486.000000   12486.000000  12486.000000   
mean             15772.862166    495.042688      46.373939     30.807144   
std               8678.364198    277.438547      57.382834     70.845471   
min                  1.000000      1.000000       1.000000      0.000000   
25%               8392.250000    286.000000       6.000000      0.000000   
50%              19228.500000    480.500000      25.000000      6.000000   
75%              23995.750000    689.000000      56.000000     28.000000   
max              27662.000000   1047.000000     213.000000    765.000000   

           position          wins  
count  12486.000000  12486.000000  
mean       7.349351      0.663143  
std        4.418334      1.773366  
min        1.000000      0.000000  
25%        4.000000      0.000000  
50%        7.000000      0.000000  
75%       10.000000      0.000000  
max       22.000000     19.000000

In [298]:
# the check the year of the last raceId in races
races.sort_values('raceId', ascending=False).head(1)

raceId  year  round  circuitId                  name        date  \
996    1009  2018     21         24  Abu Dhabi Grand Prix  2018-11-25   

         time                                                url  
996  13:00:00  http://en.wikipedia.org/wiki/2018_Abu_Dhabi_Gr...

In [284]:
# let's look at raceId higher than this in constructor_standings
constructor_standings[constructor_standings['raceId'] > 1009]

constructorStandingsId  raceId  constructorId  points  position  \
12106                   27197    1010              4     6.0         5   
12107                   27196    1010            210     8.0         4   
12108                   27195    1010              6    22.0         2   
12109                   27194    1010              9    15.0         3   
12110                   27193    1010            131    44.0         1   
...                       ...     ...            ...     ...       ...   
12481                   27658    1047              4   181.0         5   
12482                   27659    1047             51     8.0         8   
12483                   27660    1047              3     0.0        10   
12484                   27661    1047              9   319.0         2   
12485                   27662    1047            210     3.0         9   

      positionText  wins  
12106            5     0  
12107            4     0  
12108            2     0  
12109            3     0  
12110            1     1  
...            ...   ...  
12481            5     0  
12482            8     0  
12483           10     0  
12484            2     2  
12485            9     0  

[380 rows x 7 columns]

In [312]:
# check no.of raceIds 
len(constructor_standings['raceId'][constructor_standings['raceId'] > 1009].unique().tolist())

38

#### Assumption: 
the extra raceIds are records of races are for the 2019, 2020 season which happened after the date 2018-11-25, the last race for 2018 season. 

#### constructors.csv

In [316]:
constructors.head(5)

constructorId constructorRef        name nationality  \
0              1        mclaren     McLaren     British   
1              2     bmw_sauber  BMW Sauber      German   
2              3       williams    Williams     British   
3              4        renault     Renault      French   
4              5     toro_rosso  Toro Rosso     Italian   

                                                 url  
0               http://en.wikipedia.org/wiki/McLaren  
1            http://en.wikipedia.org/wiki/BMW_Sauber  
2  http://en.wikipedia.org/wiki/Williams_Grand_Pr...  
3  http://en.wikipedia.org/wiki/Renault_in_Formul...  
4   http://en.wikipedia.org/wiki/Scuderia_Toro_Rosso

In [317]:
# given the columns, let's check if any unusual values for constructorId
constructors.describe()

constructorId
count     211.000000
mean      107.028436
std        61.637258
min         1.000000
25%        54.500000
50%       107.000000
75%       159.500000
max       213.000000

In [318]:
# check any nan exists
constructors.isna().sum()

constructorId     0
constructorRef    0
name              0
nationality       0
url               0
dtype: int64

#### circuits.csv

In [320]:
circuits.head(5)

circuitId   circuitRef                            name      location  \
0          1  albert_park  Albert Park Grand Prix Circuit     Melbourne   
1          2       sepang    Sepang International Circuit  Kuala Lumpur   
2          3      bahrain   Bahrain International Circuit        Sakhir   
3          4    catalunya  Circuit de Barcelona-Catalunya     MontmelÌ_   
4          5     istanbul                   Istanbul Park      Istanbul   

     country       lat        lng   alt  \
0  Australia -37.84970  144.96800  10.0   
1   Malaysia   2.76083  101.73800   NaN   
2    Bahrain  26.03250   50.51060   NaN   
3      Spain  41.57000    2.26111   NaN   
4     Turkey  40.95170   29.40500   NaN   

                                                 url  
0  http://en.wikipedia.org/wiki/Melbourne_Grand_P...  
1  http://en.wikipedia.org/wiki/Sepang_Internatio...  
2  http://en.wikipedia.org/wiki/Bahrain_Internati...  
3  http://en.wikipedia.org/wiki/Circuit_de_Barcel...  
4         http://en.wikipedia.org/wiki/Istanbul_Park

In [123]:
# can already see nan in alt, check how many and if any other
circuits.isna().sum()

circuitId      0
circuitRef     0
name           0
location       0
country        0
lat            0
lng            0
alt           72
url            0
dtype: int64

In [325]:
circuits.describe()

circuitId        lat         lng   alt
count  73.000000  73.000000   73.000000   1.0
mean   37.000000  33.872365    1.722826  10.0
std    21.217131  23.385972   65.374188   NaN
min     1.000000 -37.849700 -118.189000  10.0
25%    19.000000  33.578600   -9.394170  10.0
50%    37.000000  41.366400    3.930830  10.0
75%    55.000000  47.219700   14.764700  10.0
max    73.000000  57.265300  144.968000  10.0

only column alt has nan, this should also not affect analysis as per the tasks

#### status.csv

In [323]:
status.head(5)

statusId        status
0         1      Finished
1         2  Disqualified
2         3      Accident
3         4     Collision
4         5        Engine

In [124]:
# check if any nan exists
status.isna().sum()

statusId    0
status      0
dtype: int64

status is a simple table which just records the different status.

#### pitStops.csv

In [324]:
pitStops.head(5)

raceId  driverId  stop  lap      time duration  milliseconds
0     841       153     1    1  17:05:23   26.898         26898
1     841        30     1    1  17:05:52   25.021         25021
2     841        17     1   11  17:20:48   23.426         23426
3     841         4     1   12  17:22:34   23.251         23251
4     841        13     1   13  17:24:10   23.842         23842

In [326]:
# check if any nan exists
pitStops.isna().sum()

raceId          0
driverId        0
stop            0
lap             0
time            0
duration        0
milliseconds    0
dtype: int64

In [327]:
pitStops.describe()

raceId     driverId         stop          lap  milliseconds
count  6251.000000  6251.000000  6251.000000  6251.000000  6.251000e+03
mean    907.860342   429.468245     1.836346    24.813470  4.774453e+04
std      44.327698   397.079565     0.935314    14.150022  1.791970e+05
min     841.000000     1.000000     1.000000     1.000000  1.289700e+04
25%     868.000000    13.000000     1.000000    13.000000  2.176800e+04
50%     902.000000   807.000000     2.000000    24.000000  2.334000e+04
75%     950.000000   821.000000     2.000000    35.000000  2.554700e+04
max     988.000000   843.000000     6.000000    74.000000  2.011266e+06

In [330]:
# let's check on the time column, make sure the values are reasonable
pitStops['time'].min(), pitStops['time'].max()

('13:04:31', '21:53:08')

#### lapTimes.csv

In [331]:
lapTimes.head(5)

raceId  driverId  lap  position      time  milliseconds
0     841        20    1         1  1:38.109         98109
1     841        20    2         1  1:33.006         93006
2     841        20    3         1  1:32.713         92713
3     841        20    4         1  1:32.803         92803
4     841        20    5         1  1:32.342         92342

In [332]:
lapTimes.describe()

raceId       driverId            lap       position  \
count  426633.000000  426633.000000  426633.000000  426633.000000   
mean      423.120853     186.503817      29.827172       9.646896   
std       381.542422     316.123374      18.402245       5.580664   
min         1.000000       1.000000       1.000000       1.000000   
25%       100.000000      14.000000      14.000000       5.000000   
50%       205.000000      26.000000      29.000000       9.000000   
75%       881.000000      71.000000      44.000000      14.000000   
max       988.000000     843.000000      78.000000      24.000000   

       milliseconds  
count  4.266330e+05  
mean   9.580222e+04  
std    7.475541e+04  
min    6.741100e+04  
25%    8.238200e+04  
50%    9.080000e+04  
75%    1.027380e+05  
max    7.507547e+06

In [335]:
# let's check on the time column, just like with pitStops
lapTimes['time'].min(), lapTimes['time'].max()

('10:32.179', '9:45.712')

In [127]:
# check if any nan exists
lapTimes.isna().sum()

raceId          0
driverId        0
lap             0
position        0
time            0
milliseconds    0
dtype: int64